File: leetcode/leetcode_py/word-search-ii.py
from typing import List
'''
https://leetcode.com/problems/word-search-ii

// Solution1. Backtracking(DSF) with Trie
// time: o( m ( 4*3^(L-1) )). m(board) L(length of word)
// space: o(w + L). w(words in tires) + L(length of word. recursion)
'''
class Solution1:
    def find(self, board, trie, r, c, word, res):
        if r<0 or c<0 or r>=len(board) or c>=len(board[0]) or board[r][c] == '*' or not board[r][c] in trie:
            return
        
        temp = board[r][c]
        board[r][c] = '*'
        
        trie = trie[temp]
        word += temp
        if ('#' in trie):
            res.append(word)
            del trie['#']
            
        offset = [0,1,0,-1,0]
        for i in range(4):
            self.find(board, trie, r+offset[i], c+offset[i+1], word, res)
        board[r][c] = temp
        
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        #Build trie
        trie = {}
        for w in words:
            t = trie
            for c in w:
                if c not in t:
                    t[c] = {}
                t=t[c]
            t['#'] = '#'
        res = []
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                self.find(board, trie, i, j, '', res)
                
        return res

class Trie:
    def __init__(self):
        self.isWord = False
        self.children = {}
    def addWord(self, word):
        cur = self
        for c in word:
            if c not in cur.children: cur.children[c] = Trie()
            cur = cur.children[c]
        cur.isWord = True

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        # build trie
        root = Trie()
        for word in words: root.addWord(word)

        res = []
        def dfs(i, j, node, cur):
            if i < 0 or j < 0 or i>= len(board) or j>=len(board[0]) or board[i][j] == '#': return
            t = board[i][j]
            if t not in node.children: return

            cur += t
            node = node.children[t]
            if node.isWord:
                res.append(cur)
                node.isWord = False

            board[i][j] = '#'

            dfs(i-1, j, node, cur)
            dfs(i, j-1, node, cur)
            dfs(i+1, j, node, cur)
            dfs(i, j+1, node, cur)

            board[i][j] = t
            

        for i in range(len(board)):
            for j in range(len(board[0])):
                dfs(i, j, root, "")
                
        return res
        

File: leetcode/leetcode_cpp/word-search-ii.cpp
#include <iostream>
#include <vector>
#include <cassert>

using namespace std;

/**
https://leetcode.com/problems/word-search-ii

// Solution1. Backtracking(DSF) with Trie
// time: o( m ( 4*3^(L-1) )). m(board) L(length of word)
// space: o(w + L). w(words in tires) + L(length of word. recursion)

- backtracking: change, recursion, undo
- trie hold dictionary of words.
time: insert/search o(n)
space: o(W*L) W(number of word) L(length of word)

1. build tries from words
2. backtracking on board to find words(tries)
3. mark as visited and unmark after

*/

class Trie {
public:
    Trie() { m = vector<Trie*>(26, nullptr); }
    ~Trie() {
        for (auto v : m) delete v;
        m.clear();
    }

    void addWord(string& word) {
        auto cur = this;
        for (auto c:word) {
            int idx = c - 'a';
            if (!cur->m[idx]) cur->m[idx] = new Trie();
            cur = cur->m[idx];
        }
        cur->isWord = true;
    }
    vector<Trie*> m;
    bool isWord = false;
};

class Solution {
public:
    void dfs(vector<vector<char>>& board, int i, int j, Trie* node, string& cur, vector<string>& res) {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] == '*') return;
        node = node->m[board[i][j] - 'a'];
        if (!node) return;

        auto t = board[i][j];
        cur.push_back(t);

        board[i][j] = '*';
        if (node->isWord) {
            res.push_back(cur);
            node->isWord = false;
        }
        dfs(board, i - 1, j, node, cur, res);
        dfs(board, i, j - 1, node, cur, res);
        dfs(board, i + 1, j, node, cur, res);
        dfs(board, i, j + 1, node, cur, res);

        board[i][j] = t;
        cur.pop_back();
    }

    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        Trie* root = new Trie();
        for (auto& word: words) {
            root->addWord(word);
        }
        vector<string> res;
        string cur;
        for (int i=0; i<board.size(); ++i) {
            for (int j=0; j<board[0].size(); ++j) {
                dfs(board, i, j, root, cur, res);
            }
        }
        return res;
    }
};

int main()
{
    auto input1_1 = vector<vector<char>> {{'o','a','a','n'}, {'e','t','a','e'}, {'i','h','k','r'}, {'i','f','l','v'}};
    auto input1_2 = vector<string> {"oath","pea","eat","rain"};
    auto output = Solution().findWords(input1_1, input1_2);
    assert(find(output.begin(), output.end(), "eat") != output.end());
    assert(find(output.begin(), output.end(), "oath") != output.end());
    return 0;
}